### PR TITLE
Waiting for resolving promise on client

### PR DESCRIPTION
**Pull request by toxahak**
Added waiting for resolving promise in case of client rendering.
Some bug fixes and refactoring.

### CODE REVIEW

File: modules/ReduxAsyncConnect.js [hunk 1]
```diff
@@ -51,9 +51,10 @@ class ReduxAsyncConnect extends React.Component {
   static propTypes = {
     components: array.isRequired,
     params: object.isRequired,
+    location: object.isRequired,
     render: func.isRequired,
-    toggleLoading: func.isRequired,
-    initialDataLoaded: func.isRequired,
+    beginLoad: func.isRequired,
```
Comment (sars) [inline modules/ReduxAsyncConnect.js]: it would be great to rename it to smth like beginGlobalLoad

File: modules/ReduxAsyncConnect.js [hunk 2]
```diff
@@ -92,28 +93,21 @@ class ReduxAsyncConnect extends React.Component {
...
-      this.setState({ prevProps: this.props, loading: true });
-      this.props.toggleLoading(true);
-
-      return Promise.all(promises).catch(error => console.error('reduxAsyncConnect server promise error: ' + error))
-        .then(() => {
-            this.setState({ prevProps: null, loading: false });
-          this.props.toggleLoading(false);
-        });
-      }
+    this.props.beginLoad();
```
Comment (sars) [inline modules/ReduxAsyncConnect.js]: revert if (promises.length !== 0) { 
please ...

File: modules/asyncConnect.js [hunk 3]
```diff
@@ -4,22 +4,24 @@ const LOAD = 'reduxAsyncConnect/LOAD';
...
   const stateSlice = state[action.key];
 
   switch (action.type) {
-    case INITIAL_DATA_LOADED:
+    case BEGIN_LOAD:
       return {
         ...state,
-        loaded: true
+        loading: true,
+        loaded: false
```
Comment (sars) [inline modules/asyncConnect.js:17]: remove loading prop, because it completely reflexes !loaded

File: modules/ReduxAsyncConnect.js [hunk 4]
```diff
@@ -71,49 +72,44 @@ class ReduxAsyncConnect extends React.Component {
...
   componentWillReceiveProps(nextProps) {
-    const { components, params } = nextProps;
-    this.loadAsyncData(components, params);
+    this.loadAsyncData(nextProps);
   }
 
-  loadAsyncData(components, params) {
-    const helpers = this.props.helpers;
+  loadAsyncData(props) {
+    const { components, params } = props;
```
Comment (sars) [inline modules/ReduxAsyncConnect.js]: const { components, params, helpers } = props;

**Comment by sars** [comment]
thanks, @toxahak , cool!

File: modules/ReduxAsyncConnect.js [hunk 5]
```diff
@@ -43,7 +43,7 @@ function asyncConnectPromises(components, params, store, helpers) {
 export function loadOnServer({ components, params }, store, helpers) {
   return Promise.all(asyncConnectPromises(filterAndFlattenComponents(components), params, store, helpers))
     .catch(error => console.error('reduxAsyncConnect server promise error: ' + error)).then(() => {
-      store.dispatch(initialDataLoaded());
+      store.dispatch(endLoad());
```
Comment (YAndrew91) [inline modules/ReduxAsyncConnect.js:46]: @sars, `endLoad` isn't defined anywhere, should be `endGlobalLoad`

Comment (sars) [inline modules/ReduxAsyncConnect.js:46]: @YAndrew91 thanks, fixed. Going to write tests to avoid such problems in future.

**Pull request merged**